pkg/controller/core.oam.dev/v1alpha2/appdeployment/appdeployment_controller.go

```diff
@@ -0,0 +1,561 @@
+	}
+}
+
+func convertRawExtention2AppConfig(raw runtime.RawExtension) (*oamcorealpha.ApplicationConfiguration, error) {
+	obj := &oamcorealpha.ApplicationConfiguration{}
+	b, err := raw.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(b, obj); err != nil {
+		return nil, err
+	}
+	return obj, nil
+}
+
+func convertRawExtention2Component(raw runtime.RawExtension) (*oamcorealpha.Component, error) {
+	obj := &oamcorealpha.Component{}
+	b, err := raw.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(b, obj); err != nil {
+		return nil, err
+	}
+	return obj, nil
+}
+
+func (r *Reconciler) calculateDiff(appd *oamcore.AppDeployment) *revisionsDiff {
+	d := &revisionsDiff{}
+
+	// Note: use (AC, cluster) as the key.
+	curDict := make(map[revision]int)
+	targetDict := make(map[revision]struct{})
+
+	target := appd.Spec.AppRevisions
+
+	for _, p := range appd.Status.Placement {
+
+		for _, c := range p.Clusters {
+			key := revision{
+				RevisionName: p.RevisionName,
+				ClusterName:  c.ClusterName,
+			}
+			curDict[key] = c.Replicas
+		}
+	}
+
+	for _, rev := range target {
+		for _, p := range rev.Placement {
+			clusterName := ""
+			if p.ClusterSelector != nil {
+				clusterName = p.ClusterSelector.Name
+			}
+			key := revision{
+				RevisionName: rev.RevisionName,
+				ClusterName:  clusterName,
+			}
+			targetDict[key] = struct{}{}
+
+			curReplicas, ok := curDict[key]
+
+			toAdd := newRevision(rev.RevisionName, clusterName, p.Distribution.Replicas)
+			if !ok {
+				// need to add
+				d.Add = append(d.Add, toAdd)
+				continue
+			}
+
+			if p.Distribution.Replicas == curReplicas {
+				d.Unchanged = append(d.Unchanged, toAdd)
+				continue
+			}
+			// need to mod
+			d.Mod = append(d.Mod, toAdd)
+		}
+	}
+
+	for _, p := range appd.Status.Placement {
+		for _, c := range p.Clusters {
+			key := revision{
+				RevisionName: p.RevisionName,
+				ClusterName:  c.ClusterName,
+			}
+
+			_, ok := targetDict[key]
+			if ok {
+				continue
+			}
+			// need to del
+			d.Del = append(d.Del, newRevision(p.RevisionName, c.ClusterName, c.Replicas))
+		}
+	}
+	return d
+}
+
+// UpdateStatus updates AppDeployment's Status with retry.RetryOnConflict
+func (r *Reconciler) updateStatus(ctx context.Context, appd *oamcore.AppDeployment, opts ...client.UpdateOption) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() (err error) {
+		return r.Client.Status().Update(ctx, appd, opts...)
+	})
+}
+
+// SetupWithManager setup the controller with manager
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&oamcore.AppDeployment{}).
+		Complete(r)
+}
+
+func (r *Reconciler) applyTraffic(ctx context.Context, appd *oamcore.AppDeployment) (err error) {
+
+	vsvc := &istioclientv1beta1.VirtualService{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: istioclientv1beta1.SchemeGroupVersion.String(),
+			Kind:       "VirtualService",
```

Comment:
Do we have any extensibility for other service mesh implementation here? I means, if we wan to support e.g. OSM, how to extend the code.

Reply:
Yes. I think there should be a service mesh interface here to manage traffic. This will need more user input to design a generic interface which we should not over-optimize for now.